### PR TITLE
ci: add new action for spec release

### PR DESCRIPTION
**Pull request by aayushmau5**
---
title: "New GH Action"
---
This is a new github action which will run whenever a `release` event is fired(On spec release). It will copy all the necessary spec files into the website repository and make a PR on website repository.

**Related issue(s):**
---
https://github.com/asyncapi/website/issues/86

### CODE REVIEW

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:32]: instead of doing this you can do 
```
working-directory: ./website
```

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:46]: instead of doing this you can do 
```
working-directory: ./website
```

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:36]: there is no such prefix as `spec` in conventional commits specification

```suggestion
          git commit -m "docs(spec): ${{github.event.release.tag_name}} release"
```

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:47]: there is no such prefix as `spec` in conventional commits specification
```suggestion
          gh pr create --title "docs(spec): ${{github.event.release.tag_name}} release" --body "New version of the specification is available and this PR introduces update to specification document on the website" --head "${{github.event.release.tag_name}}"
```

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:38]: using action from `master` is very risky and vulnerable, please use `v0.6.0` version

anyway do we need this? can't you just do `git push https://${{ env.GITHUB_TOKEN }}@github.com/asyncapi/website` in the previous step?

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:1]:
```suggestion
name: Update spec document because of new spec release
```

**Review by derberg — CHANGES_REQUESTED** [review]
@aayushmau5 hey, I left some comments. Also have 2 questions:
- did you test the workflow somewhere on some other repos to make sure it actually works?
- the approach with copy paste is risky as the spec file name is `asyncapi.md` so for each version it will be the same, so if the workflow runs on a release branch, then the spec on website will be updated with version from release branch. I guess this will clarify once you jump into the website repo and start working on this there too

**Comment by aayushmau5** [comment]
Hey @derberg Thanks for the review. I will make the requested changes. 
>  did you test the workflow somewhere on some other repos to make sure it actually works?

Yup, I did try it in one of my repos. Though I have since deleted it. I will make another for your verification & check if the changes work or not.

> the approach with copy paste is risky as the spec file name is asyncapi.md so for each version it will be the same, so if the workflow runs on a release branch, then the spec on website will be updated with version from release branch. I guess this will clarify once you jump into the website repo and start working on this there too

You are right. :thinking: perhaps we prefix file names with their release version? or make directories?

**Comment by derberg** [comment]
> You are right. 🤔 perhaps we prefix file names with their release version? or make directories with release name and put the spec files inside them? You got any other ideas?

yeah, I think the idea to keep the `asyncapi.md` name but just place it in the dedicated folder with the version name is good. 
Something that you have to confirm when working on the website.

File: .github/workflows/new-spec-release.yml
Comment (aayushmau5) [reply .github/workflows/new-spec-release.yml]: Changed it to `it push https://${{ env.GITHUB_TOKEN }}@github.com/asyncapi/website`

File: .github/workflows/new-spec-release.yml
Comment (aayushmau5) [reply .github/workflows/new-spec-release.yml:25]: Yeah, so about that. In my own repository, I was getting this permission error. 
Screenshot_2021-05-26_11 35 04

Add adding `token: ${{ env.GITHUB_TOKEN }}` fixed that. I'm not sure if this error will occur in `spec` repo or not.

**Comment by aayushmau5** [comment]
@derberg I setup repositories for preview. Check it out here

1. https://github.com/aayushmau5/demo-gh-action-test-1 (Equivalent to `spec`)
  - Action ran when `1.2` is released https://github.com/aayushmau5/demo-gh-action-test-1/actions/runs/877466859
2. https://github.com/aayushmau5/demo-gh-action-test-2 (Equivalent to `website`)
  - PR link generated by Action https://github.com/aayushmau5/demo-gh-action-test-2/pulls

These were for the most basic release cases. Do let me know if we need to make some more tests.(Like pre-release etc.)

File: .github/workflows/new-spec-release.yml
Comment (derberg) [reply .github/workflows/new-spec-release.yml:25]: it most probably will, I guess token is needed by default when cloning, I actually do the same in one of my actions, this way without code complications I support both private and public repos.

just leave it as it is then

**Comment by derberg** [comment]
@aayushmau5 that looks perfect so far, great work!

> Do let me know if we need to make some more tests.(Like pre-release etc.)

there should be no difference really. Pre-release like any other, will also trigger this workflow.
I think that even patch releases can be treated individually here. Once we get to many, we can think on some improvement, right?

we only need one enhancement here, the release that is marked as `latest` should be copied twice, into folder with the name of the release and folder called `latest`. In the website UI, whatever is in `latest` folder would be treated as latest and always displayed as default. Does it make sense? so for example pre-releases are not marked as latest, so when someone will go to website UI to read the spec, by default they will not see release candidate and won't be confused to see things that are actually not yet in spec.

**Comment by aayushmau5** [comment]
> we only need one enhancement here, the release that is marked as latest should be copied twice, into folder with the name of the release and folder called latest. In the website UI, whatever is in latest folder would be treated as latest and always displayed as default. Does it make sense? so for example pre-releases are not marked as latest, so when someone will go to website UI to read the spec, by default they will not see release candidate and won't be confused to see things that are actually not yet in spec.

Yeah, that makes perfect sense. I will starting working on it :)

**Comment by aayushmau5** [comment]
@derberg I have added the support for `latest` directory. Please review. 

For real life example, see the Action at https://github.com/aayushmau5/demo-gh-action-test-1/actions
and PR at https://github.com/aayushmau5/demo-gh-action-test-2/pulls

**Review by derberg — DISMISSED** [review]
Awesome stuff! I approve but will merge once we have the proper support on the website. Will you do this part too?

**Comment by aayushmau5** [comment]
Yeah, I'm gonna start working on website as well.

**Comment by aayushmau5** [comment]
@derberg @magicmatatjahu I have now added the functionality to automatically delete the Pre-release file on `release` as well as modify the `_redirects` to point to the latest release file. Please review.

For real example, Here's my "testing" repo
1. https://github.com/aayushmau5/actions-test-1 
    Actions: https://github.com/aayushmau5/actions-test-1/actions

2. https://github.com/aayushmau5/actions-test-2
   PR made by action: https://github.com/aayushmau5/actions-test-2/pull/2

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:44]: no need to append `v`. Previous releases didn't have it, but next ones will, I even added v2.0.0 tag recently,

File: .github/workflows/new-spec-release.yml
Comment (derberg) [inline .github/workflows/new-spec-release.yml:50]: there can be more than one prereleases there. I mean you can have `v2.1.0-2021-06-release.1`, `v2.1.0-2021-06-release.2` etc, so you will have to modify `matchedFileName` to array of filenames and unlink multiple

**Review by derberg — COMMENTED** [review]
I love how you figured to cleanup prereleases! clean solution

I left 2 comments 👏🏼

File: .github/workflows/new-spec-release.yml
Comment (aayushmau5) [reply .github/workflows/new-spec-release.yml]: Cleaned it up a bit.

**Comment by derberg** [comment]
looks perfect, now just switch from GITHUB_TOKEN to GH_TOKEN so all changes are done not by default `github actions` bot but our `asyncapi bot`

**Comment by aayushmau5** [comment]
@derberg Are you talking about this line?
```
env:
    GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
```

**Comment by derberg** [comment]
@aayushmau5 ah, sorry, forgot you take GH_TOKEN once into the env, and then reuse it, forget about my comment then

**Review by derberg — APPROVED** [review]
Well done 🚀 
Can be merged once all changes on the website are introduced

**Comment by aayushmau5** [comment]
Need to make some changes before Merging.

Ref: https://github.com/asyncapi/website/pull/276

**Comment by aayushmau5** [comment]
@derberg Added the ability to remove the `# AsyncAPI Specification` heading and modify the `# SPEC-REDIRECTION` as well as `# LATEST-SPEC-REDIRECTION` line.

Please review.

**Comment by derberg** [comment]
@aayushmau5 it is looking great. No remarks, perfect work!

Is it all? I have to admit that after my short holidays reset I lost track of things 😅

**Comment by aayushmau5** [comment]
Yeah, I think so. I looked at the issue to see if there's anything I have missed, and no, I don't think I did :D

**Review by derberg — APPROVED** [review]
LGTM 🚀 
Thanks a lot @aayushmau5 for the huge effort!

**Pull request merged**
